test/core/rules/test_cases/L004.yml

```diff
@@ -1,10 +1,106 @@
 rule: L004
 
-test_1:
-  pass_str: "   \nSELECT 1"
+spaces_pass_default:
+  pass_str: "    \nSELECT 1"
 
-test_2:
-  pass_str: "\t\tSELECT 1\n"
 
-test_3:
-  fail_str: "   \n  \t \n  SELECT 1"
+spaces_fail:
+  fail_str: "    \nSELECT 1"
+  fix_str: "\t\nSELECT 1"
+
+  configs:
+    rules:
+      indent_unit: tab
+
+
+spaces_fail_custom_tab_space_size:
+  fail_str: "    \nSELECT 1"
+  fix_str: "\t\t\nSELECT 1"
+
+  configs:
+    rules:
+      indent_unit: tab
+      tab_space_size: 2
+
+
+tabs_fail_default:
+  fail_str: "\t\tSELECT 1\n"
+  fix_str: "        SELECT 1\n"
+
+
+tabs_fail_default_set_tab_space_size:
+  fail_str: "\t\tSELECT 1\n"
+  fix_str: "    SELECT 1\n"
+
+  configs:
+    rules:
+      tab_space_size: 2
+
+
+tabs_pass:
+  pass_str: "\tSELECT 1"
+
+  configs:
+    rules:
+      indent_unit: tab
+
+
+mixed_indent_fail_default_tab_space_size:
+  fail_str: |2
+      	 select 1
+  fix_str: |2
+           select 1
+
+
+mixed_indent_fail_custom_tab_space_size:
+  fail_str: |2
+      	 select 1
+  fix_str: |2
+         select 1
+
+  configs:
+    rules:
+      tab_space_size: 2
+
+
+indented_comments:
+  pass_str: |
+    SELECT
+        a,         -- Some comment
+        longer_col -- A lined up comment
+    FROM spam
+
+indented_comments_default_config:
+  fail_str: |
+    SELECT
+    	a,			-- Some comment
+    	longer_col	-- A lined up comment
+    FROM spam
+
+  # The rule will only fix the indent before the select targets.
+  fix_str: |
+    SELECT
+        a,			-- Some comment
```

Comment:
test demonstrating that the tab is replaced with spaces only after the newline, not after the text.